types/react/index.d.ts

```diff
@@ -79,12 +79,15 @@ declare namespace React {
         | ((props: P) => ReactElement<any, any> | null)
         | (new (props: P) => Component<any, any>);
 
-    interface RefObject<T> {
-        readonly current: T | null;
-    }
+    /** 
+     * @deprecated All refs are mutable. This type has always been wrong. Expect it to change to
```

Comment:
This wording doesn't really add value. There [was a rationale why we did this](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useref). It's like saying `ReadonlyArray<SomeArray>` "was always wrong". Just because something works in JavaScript, doesn't mean it's correct.

I don't think we should deprecate this. Doesn't it make more sense to drop MutableRefObject in the future instead and let `RefObject` be the mutable one?

Reply:
Originally, MutableRefs meant user-managed while readonly-refs meant react-managed. See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64772/files#r1148507326

Reply:
I reviewed the history prior to making the recommendation. However there isn't really a way to know if the intended use is user managed vs react-managed, which is unfortunate. 

Changing RefObject to be representable with MutableRefObject allows for deprecation. The reverse isn't possible.

I'm saying its "wrong" because really you can change it as you see fit and applying magic typing to try to signify if something is readonly or not is just confusing. I'm saying its "wrong" because we created a weird mechanism in the types that are inconsistent with react itself and have been since day one. 

Also, saying the ref is 'readonly' when you pass it to something that clearly modifies the value, also breaks most of the ways that readonly is used, which is that it lets you know that the thing you are giving it to doesn't mutate it. To your example of a ReadonlyArray, if an API had a `doFoo()` that returned `{foo: ReadonlyArray<T>}` then I would assume that the inner array would never change out from under me. But in the analogy, you're allowing it to change by react.